Optimize StyleOriginatedAnimation::cancel()

NathanSolomon733 opened the pull request:
#### aa4a33e6bea441f0f89dab3926e2edadb0471c64
<pre>
Optimize StyleOriginatedAnimation::cancel()

<a href="https://bugs.webkit.org/show_bug.cgi?id=303456">https://bugs.webkit.org/show_bug.cgi?id=303456</a>

<a href="https://rdar.apple.com/165741697">rdar://165741697</a>

Reviewed by Antoine Quint.

StyleOriginatedAnimation::cancel is a hot path in SP3.
To optimize this function for cancellation events, skip evaluating the
conditions to decide how to enqueue the DOM event,
since we already know the parameters enqueueDOMEvent
needs. Most importantly, we do not need to call getComputedTiming()
to enqueue the cancel event. getComputedTiming() is the most
expensive part of calling invalidateDOMEvents, so skip this step.

A/B testing shows this is a performance improvement.

* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::setTimeline):
(WebCore::StyleOriginatedAnimation::cancel):
(WebCore::StyleOriginatedAnimation::computeCancelationTime const):
* Source/WebCore/animation/StyleOriginatedAnimation.h:

Canonical link: <a href="https://commits.webkit.org/304895@main">https://commits.webkit.org/304895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37d4ffb591466f74ddf0a8732c74b1be69022c34

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48013 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144457 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89699 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c29ce98d-b6d2-483b-8d17-48f04d2028fe) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138598 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8931 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104552 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/105cdde0-0d26-42b5-9437-7b71a508df6b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7154 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85391 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/664dfc19-4871-491f-a176-5a94ef51e23b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6798 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4505 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5046 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40695 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147214 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41268 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112907 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8787 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7377 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113237 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28781 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6720 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118798 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62921 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8817 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36852 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8538 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72383 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8757 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8609 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/StyleOriginatedAnimation.cpp:
```diff
@@ -250,16 +269,20 @@ WebAnimationTime StyleOriginatedAnimation::effectTimeAtEnd() const
+    return 0_s;
+}
+
+template<typename F> void StyleOriginatedAnimation::invalidateDOMEvents(F&& callback)
```
This function now has a single caller. I suggest we remove this callback-based version and simply call `invalidateDOMEvents(WebAnimationTime)` from and pass in the cancelation time which we'd compute with `computeCancelationTime()`.

NathanSolomon733 replied:
Done.

graouts commented on Source/WebCore/animation/StyleOriginatedAnimation.cpp:
```diff
@@ -194,14 +194,33 @@ void StyleOriginatedAnimation::setTimeline(RefPtr<AnimationTimeline>&& newTimeli
-    invalidateDOMEvents([protectedThis = Ref { *this }, silently] {
-        protectedThis->WebAnimation::cancel(silently);
-    });
+    auto cancelationTime = computeCancelationTime();
```
You should either:

- move this to within the `if` statement below since it's only needed there,
- or leave it there, add `ASSERT(!wasIdle && !wasAfter)` and remove the `if` statement altogether (I have no idea whether that assertion would hold but I think the layout tests would indicate that with crashes in debug mode).

NathanSolomon733 replied:
The ASSERT would be ideal, but I just tried on debug mode and there's a few tests hitting the assert and then crashing, so I have to do the first option.

NathanSolomon733 replied:
Actually, we can not place `auto cancelationTime = computeCancelationTime();` in the if statement where the result is used because we would end up computing `cancelationTime` _after_ calling `Ref { *this }->WebAnimation::cancel(silently);`. WebAnimation::cancel has a side effect of setting m_holdTime and m_startTime to nullopt, which then makes the localTime nullopt, and therefore computeCancelationTime() will always return 0_s instead of the actual elapsed time.  This causes a test failure in imported/w3c/web-platform-tests/css/css-animations/event-dispatch.tentative.html.

So, I have to leave computeCancelationTime to be called unconditionally, although a bit wasteful for the cases where we return early.

graouts replied:
It's alright since `computeCancelationTime()` will just return `0_s` if `!m_owningElement`.

NathanSolomon733 committed:
Optimize StyleOriginatedAnimation::cancel()

https://bugs.webkit.org/show_bug.cgi?id=303456

rdar://165741697

Reviewed by Antoine Quint.

StyleOriginatedAnimation::cancel is a hot path in SP3.
To optimize this function for cancellation events, skip evaluating the
conditions to decide how to enqueue the DOM event,
since we already know the parameters enqueueDOMEvent
needs. Most importantly, we do not need to call getComputedTiming()
to enqueue the cancel event. getComputedTiming() is the most
expensive part of calling invalidateDOMEvents, so skip this step.

A/B testing shows this is a performance improvement.

* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::setTimeline):
(WebCore::StyleOriginatedAnimation::cancel):
(WebCore::StyleOriginatedAnimation::computeCancelationTime const):
* Source/WebCore/animation/StyleOriginatedAnimation.h:

Canonical link: https://commits.webkit.org/304895@main
